rfcs/0011-ban-scoring.md

```diff
@@ -0,0 +1,148 @@
+
+For a score of 20:
+
+- too many unconnecting headers
+- too-big message "addr", "inventory", or "getdata" message sizes
+- too many headers, non-continuous headers
+
+And a score of of 1:
+
+ - missing "version" or "verack" messages, duplicate "version" message
+
+In several places in the code, the "DoS" value is checked, and if
+positive, it's added to the ban score.  The DoS score can be
+incremented when a transaction is rejected. Many reasons for rejecting
+a transaction add 100 to the DoS value (too many to enumerate
+here). Those include items like transaction fees out of range, missing
+or already-spent inputs, incorrect proof of work, and an invalid
+Merkle root.
+
+Some reasons for rejecting a transaction increment the DoS value a
+lesser amount. For instance, and invalid hash such that the proof of
+work fails, increments DoS by 50. A previous-block-not-found rejection
+increments DoS by 10. In several cases, transactions are rejected, but
+the DoS is not incremented, such as the "mempool-full" condition.
+
+# Coda at the moment
+
+Bitcoin is a mature codebase, so there are many places where ban scoring has been
+used. Nonetheless, Bitcoin uses a relatively coarse ban scoring system; only
+a few ban score increment values are used. In Coda, we could reify scores into a
+datatype:
+
+```ocaml
+  module Ban_score = struct
+    type t =
+        | Severe
+	| Moderate
+	| Trivial
+  end
+```
+A slightly finer gradation could be used, if desired. The blacklisting system
+could translate these constructors into numerical scores. Let's call these
+constructors SEV, MOD, and TRV.
+
+In a number of places, the Coda codebase has comments indicating that
+a peer should be punished, either via a `TODO` or call to `Logger.faulty_peer`.
+Let's classify those places where punishment has been flagged, and annotate
+them with suggested constructors:
+
+- in `bootstrap_controller.ml`, for bad proofs (SEV), and a validation error when
+    building a breadcrumb (SEV)
+- in `coda_networking.ml`, when an invalid staged ledger hash is received (SEV), or
+    when a transition sender does not return ancestors (MOD)
+- in `ledger_catchup.ml`, when a root hash can't be found (SEV), or a peer returns an empty list
+    of transitions (instead of `None`) (TRV)
+- in `linked_tree.ml`, for peers requesting nonexistent ancestor paths (MOD)
+- in `parallel_scan.ml`, in `update_new_job` for unneeded merges (?) (SEV)
+- in `staged_ledger.ml`, when a bad signature is encountered when applying a pre-diff (SEV)
+- in `syncable_ledger.ml`, in `num_accounts`, when a content hash doesn't match a stored root hash (SEV),
+    and in `main_loop`, when a child hash can't be added (MOD)
+- in `catchup_scheduler.ml` and `processor.ml`, when a breadcrumb can't be built from a
+    transition (SEV) (same failure as in `bootstrap_controller.ml`, above)
+- in `ledger_catchup.ml`, a transition could not be validated (SEV)
+- in `transaction_pool.ml`, a payment check fails (SEV)
```

Comment:
also missing from here is the bit that Echo is adding about sending a transaction that doesn't make sense on the current staged ledger.

This is the bit that justifies the trust scoring I think. I think Echo was mentioning this somewhere, but if you receive transactions at high-velocity from some node and 99% of them work on the staged ledger, but some of them don't due to inconsistent state between the two nodes, we'd want that to be okay.

Whereas if you get one transaction each day from a node, maybe on the second or third wrong one, you'd want to ban it.

This is something unique to Coda w.r.t. Bitcoin because we expect to support high-transaction throughput.

However, on the one hand, you could argue that we should be banning based on how much pain or wasted work was inflicted by the accused. In that case, you'd want to ban if you ever receive some constant amount of failures from a node, despite seeing correct ones and no matter in what rate you're receiving them.

Reply:
added section about trust scoring